### PR TITLE
language_model optimizations

### PR DESCRIPTION
**Pull request by smassung**
This improves the `language_model` operation running times. As a test on my own computer, these changes made `diff-test` run in 1m12s compared to 1m55s originally on the CoNLL data.

### CODE REVIEW

File: include/lm/token_list.h [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+template <class HashAlgorithm>
+void hash_append(HashAlgorithm& h, const token_list& list)
+{
+    using util::hash_append;
+    for (const auto& val : list.tokens())
+        hash_append(h, val);
+    hash_append(h, list.size());
+}
```
Comment (skystrife) [inline include/lm/token_list.h]: Potentially faster (probably doesn't matter in practice):

```cpp
template <class HashAlgorithm>
void hash_append(HashAlgorithm& h, const token_list& list)
{
    h(tokens_.data(), sizeof(term_id) * list.size());
    hash_append(h, list.size());
}
```

(You'll need to friend your `hash_append` overload to do this, but that's easy.)

File: include/lm/language_model.h [hunk 2]
```diff
@@ -111,9 +117,11 @@ class language_model
 
     std::vector<static_probe_map> lm_;
 
-    std::vector<std::string> vocabulary_;
+    std::unordered_map<std::string, term_id> vocabulary_;
```
Comment (skystrife) [inline include/lm/language_model.h]: `util::probe_set` is another option.

File: src/lm/static_probe_map.cpp [hunk 3]
```diff
@@ -56,10 +56,10 @@ util::optional<lm_node> static_probe_map::find(const std::string& key) const
     }
 }
 
-uint64_t static_probe_map::hash(const std::string& str) const
+uint64_t static_probe_map::hash(const token_list& tokens) const
 {
     util::murmur_hash<> hasher{seed_};
-    hasher(str.data(), str.length());
+    hasher(tokens.tokens().data(), tokens.size() * sizeof(term_id));
```
Comment (skystrife) [inline src/lm/static_probe_map.cpp]: aka `util::hash_append(hasher, tokens)`

File: include/lm/token_list.h (hunk 1 above)
Comment (smassung) [inline include/lm/token_list.h]: Ok, changed to that.

**Comment by smassung** [comment]
Wow, changing the hash stuff actually made `diff-test` take 1m0s as opposed to 1m12s.

**Comment by smassung** [comment]
Nevermind, that was using the old lm files that weren't regenerated. It's about the same.

File: include/lm/token_list.h [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+    /**
+     * Remove the last token
+     */
+    void pop_back();
+
+    /**
+     * @return the underlying container of term_id tokens
+     */
+    const std::vector<term_id> tokens() const;
```
Comment (skystrife) [inline include/lm/token_list.h]: `&`. `&` very important.

I missed this until I profiled.

Comment (smassung) [inline include/lm/token_list.h]: Wow. Just fixed that.

**Pull request merged**